src/librustc/ty/context.rs

```diff
@@ -1510,6 +1508,49 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
     {
         self.mk_substs(iter::once(s).chain(t.into_iter().cloned()).map(Kind::from))
     }
+
+    pub fn lint_node<S: Into<MultiSpan>>(self,
+                                         lint: &'static Lint,
+                                         id: NodeId,
+                                         span: S,
+                                         msg: &str) {
+        self.struct_span_lint_node(lint, id, span.into(), msg).emit()
+    }
+
+    pub fn lint_level_at_node(self, lint: &'static Lint, mut id: NodeId)
+        -> (lint::Level, lint::LintSource)
+    {
+        let sets = self.lint_levels(LOCAL_CRATE);
```

Comment:
Indeed. This is where I have only one suggestion: make the lint computation bottom-up instead of top-down, i.e. the query gets its parent node's lint levels, then appends the ones from its own attributes and returns. That should have  really good tracking.